src/librustc_trans/back/write.rs

```diff
@@ -557,9 +561,15 @@ unsafe fn optimize_and_codegen(cgcx: &CodegenContext,
 
         if config.emit_obj {
             let path = output_names.with_extension(&format!("{}.o", name_extra));
-            with_codegen(tm, llmod, config.no_builtins, |cpm| {
-                write_output_file(cgcx.handler, tm, cpm, llmod, &path, llvm::ObjectFileType);
-            });
+            if !cgcx.is_like_pnacl {
+                with_codegen(tm, llmod, config.no_builtins, |cpm| {
```

Comment:
Except it's LLVM's target machine that does that. PNaCl/JS (well JS does have one, but it isn't used till post-linking) doesn't have a target machine (though PNaCl does use `armv7-none-nacl-gnu` for optimizing, using it for codegen would produce an ARM NaCl object file, which isn't compatible with PNaCl).

In fact if you take a look at the toolchain's `pnacl-clang`, you'll find it's just a python wrapper which invokes clang with arguments like `-emit-llvm`.